File: server/src/quotas/proxy.ts
// server/src/quotas/proxy.ts — 额度接口出网客户端：chatgpt.com / api.anthropic.com /
// cursor.com 在受限网络不可达，需要代理。Node 内置 fetch 的代理开关
//（NODE_USE_ENV_PROXY）只在进程启动前生效，launchd 下无法保证——因此用
// undici 的 ProxyAgent 显式构造 dispatcher（纯 JS 依赖，无版本门槛）。
// 代理地址：显式 HTTP(S)_PROXY env 优先；否则 macOS 用 `scutil --proxy` 探测系统代理。
import { execFileSync } from 'node:child_process'
import type { ProxyAgent, fetch as undiciFetchType } from 'undici'
import type { FetchLike } from './types.js'

export function nodeMajorVersion(version = process.versions.node): number {
  return Number.parseInt(version.split('.')[0] ?? '0', 10) || 0
}

export type SystemProxy = { httpsProxy: string | null; httpProxy: string | null }

/** 解析 `scutil --proxy` 输出（macOS）；非 mac / 解析失败返回空 */
export function parseScutilProxy(output: string): SystemProxy {
  const get = (key: string): string | null => {
    const m = new RegExp(`^\\s*${key}\\s*:\\s*(\\S+)`, 'm').exec(output)
    return m?.[1] ?? null
  }
  const enabled = (key: string): boolean => get(key) === '1'
  const host = (key: string): string | null => {
    const v = get(key)
    return v && v !== '0.0.0.0' ? v : null
  }
  const port = (key: string): string | null => {
    const v = get(key)
    return v && v !== '0' ? v : null
  }
  const build = (enableKey: string, hostKey: string, portKey: string): string | null => {
    if (!enabled(enableKey)) return null
    const h = host(hostKey)
    const p = port(portKey)
    return h && p ? `http://${h}:${p}` : h ? `http://${h}` : null
  }
  return {
    httpsProxy: build('HTTPSEnable', 'HTTPSProxy', 'HTTPSPort'),
    httpProxy: build('HTTPEnable', 'HTTPProxy', 'HTTPPort'),
  }
}

/** 代理地址解析：显式 env 优先，其次 macOS 系统代理；都没有返回 null（直连） */
export function resolveProxyUrl(opts: {
  env?: NodeJS.ProcessEnv
  platform?: string
  scutil?: () => string
} = {}): string | null {
  const env = opts.env ?? process.env
  const platform = opts.platform ?? process.platform
  const explicit = env['HTTPS_PROXY'] ?? env['https_proxy'] ?? env['HTTP_PROXY'] ?? env['http_proxy']
  if (explicit && explicit.trim()) return explicit.trim()
  if (platform !== 'darwin') return null
  try {
    const out = opts.scutil ? opts.scutil() : execFileSync('/usr/sbin/scutil', ['--proxy'], { encoding: 'utf8' })
    const sys = parseScutilProxy(out)
    return sys.httpsProxy ?? sys.httpProxy
  } catch {
    return null
  }
}

/**
 * 构造额度接口专用 fetch：配置了代理时走 undici ProxyAgent（运行时生效），
 * 否则退回全局 fetch 直连。undici 按需懒加载（未配代理时不引入）。
 */
export async function createQuotaFetch(opts: {
  env?: NodeJS.ProcessEnv
  platform?: string
  scutil?: () => string
} = {}): Promise<{ fetchLike: FetchLike; proxyUrl: string | null }> {
  const proxyUrl = resolveProxyUrl(opts)
  if (!proxyUrl) return { fetchLike: (url, init) => fetch(url, init), proxyUrl: null }
  const mod = (await import('undici')) as { ProxyAgent: typeof ProxyAgent; fetch: typeof undiciFetchType }
  const agent = new mod.ProxyAgent(proxyUrl)
  const fetchLike: FetchLike = async (url, init) => {
    const res = await mod.fetch(url, {
      method: init.method as 'GET' | 'POST' | undefined,
      headers: init.headers,
      body: init.body as string | undefined,
      signal: init.signal,
      dispatcher: agent,
    } as Parameters<typeof mod.fetch>[1])
    return { ok: res.ok, status: res.status, json: () => res.json() }
  }
  return { fetchLike, proxyUrl }
}

File: Sources/WattsonCore/Quotas/Proxy.swift
// Quotas/Proxy.swift — 额度接口出网代理解析：显式 HTTP(S)_PROXY env 优先；
// 否则 macOS 用 `scutil --proxy` 探测系统代理（chatgpt.com 等受限域名出网）。
import Foundation

public struct SystemProxy: Sendable, Equatable {
    public var httpsProxy: String?
    public var httpProxy: String?
}

/// 解析 `scutil --proxy` 输出（macOS）；解析失败返回空
public func parseScutilProxy(_ output: String) -> SystemProxy {
    func get(_ key: String) -> String? {
        let pattern = "^\\s*\(key)\\s*:\\s*(\\S+)"
        guard let regex = try? NSRegularExpression(pattern: pattern, options: [.anchorsMatchLines]),
              let m = regex.firstMatch(in: output, range: NSRange(output.startIndex..., in: output)),
              let r = Range(m.range(at: 1), in: output) else { return nil }
        return String(output[r])
    }
    func enabled(_ key: String) -> Bool { get(key) == "1" }
    func host(_ key: String) -> String? {
        guard let v = get(key), v != "0.0.0.0" else { return nil }
        return v
    }
    func port(_ key: String) -> String? {
        guard let v = get(key), v != "0" else { return nil }
        return v
    }
    func build(_ enableKey: String, _ hostKey: String, _ portKey: String) -> String? {
        guard enabled(enableKey) else { return nil }
        let h = host(hostKey)
        let p = port(portKey)
        if let h, let p { return "http://\(h):\(p)" }
        return h.map { "http://\($0)" }
    }
    return SystemProxy(
        httpsProxy: build("HTTPSEnable", "HTTPSProxy", "HTTPSPort"),
        httpProxy: build("HTTPEnable", "HTTPProxy", "HTTPPort"))
}

/// 运行系统 scutil（生产路径）
public func runScutilProxy() -> String {
    let p = Process()
    p.executableURL = URL(fileURLWithPath: "/usr/sbin/scutil")
    p.arguments = ["--proxy"]
    let pipe = Pipe()
    p.standardOutput = pipe
    p.standardError = Pipe()
    do {
        try p.run()
        let data = pipe.fileHandleForReading.readDataToEndOfFile()
        p.waitUntilExit()
        return String(data: data, encoding: .utf8) ?? ""
    } catch {
        return ""
    }
}

/// 代理地址解析：显式 env 优先，其次 macOS 系统代理；都没有返回 nil（直连）
public func resolveProxyUrl(env: [String: String] = [:], platform: String = "darwin",
                            scutil: (() -> String)? = nil) -> String? {
    let explicit = env["HTTPS_PROXY"] ?? env["https_proxy"] ?? env["HTTP_PROXY"] ?? env["http_proxy"]
    if let explicit, !explicit.trimmingCharacters(in: .whitespaces).isEmpty {
        return explicit.trimmingCharacters(in: .whitespaces)
    }
    guard platform == "darwin" else { return nil }
    let out = scutil?() ?? runScutilProxy()
    let sys = parseScutilProxy(out)
    return sys.httpsProxy ?? sys.httpProxy
}

/// 构造额度接口专用 fetch：配置了代理时走系统代理（URLSession connectionProxyDictionary），
/// 否则直连。Swift 原生实现无进程启动时机限制，运行时生效。
public func createQuotaFetch(env: [String: String] = [:], platform: String = "darwin",
                             scutil: (() -> String)? = nil) -> (fetchLike: FetchLike, proxyUrl: String?) {
    let proxyUrl = resolveProxyUrl(env: env, platform: platform, scutil: scutil)
    return (makeHTTPFetch(proxy: proxyUrl), proxyUrl)
}
